File: ports/py/tests/integration/test_report_hint.py
"""The internal-error 'report this' hint is appended to engine panics
(surfaced as RuntimeError) but never to validation errors (ValueError).

The panic seam ``_engine.panic_for_test`` exists only in test-bridge builds
(stripped from release wheels), mirroring engine-r's ``find_power_precancelled``.
"""

import pytest

from mcpower import _engine

REPORT_URL = "mcpower.app/report"

panic_for_test = getattr(_engine, "panic_for_test", None)


@pytest.mark.skipif(panic_for_test is None, reason="panic_for_test requires a test-bridge build")
def test_internal_panic_carries_report_hint():
    """A caught engine panic surfaces as RuntimeError with the report URL."""
    with pytest.raises(RuntimeError) as exc_info:
        panic_for_test()
    msg = str(exc_info.value)
    assert REPORT_URL in msg
    assert "port=py" in msg


def test_validation_error_has_no_report_hint():
    """An engine-layer ValueError (malformed contracts) must NOT carry the hint —
    the suffix is reserved for genuine internal failures, not user input."""
    with pytest.raises(ValueError) as exc_info:
        _engine.find_power(b"\xff\xff\xff", 100, 100, 0)
    assert REPORT_URL not in str(exc_info.value)

File: ports/r/tests/testthat/test-report-hint.R
# The internal-error "report this" hint is appended to engine panics (surfaced
# as an R error) but never to validation or cancellation errors.
#
# panic_for_test is an internal test seam (mirrors engine-py's panic_for_test);
# see crates/engine-r/src/orchestrator_bridge.rs.

test_that("a caught engine panic surfaces with the report hint", {
  err <- tryCatch(mcpower:::panic_for_test(), error = function(e) conditionMessage(e))
  expect_true(grepl("mcpower.app/report", err, fixed = TRUE))
  expect_true(grepl("port=r", err, fixed = TRUE))
})

test_that("a validation error does not carry the report hint", {
  # Malformed contracts bytes → engine-layer validation error, no hint.
  err <- tryCatch(
    mcpower:::find_power(as.raw(c(0xff, 0xff, 0xff)), 100L, 100L, 2137, NULL),
    error = function(e) conditionMessage(e)
  )
  expect_false(grepl("mcpower.app/report", err, fixed = TRUE))
})

test_that("a cancellation error does not carry the report hint", {
  cj <- mcpower:::build_contract_from_spec(.ols_spec_json("y ~ x1", x1 = 0.3),
                                           "continuous", "canonical", "ols", 0.0, "[]")$contracts
  err <- tryCatch(
    mcpower:::find_power_precancelled(cj, 200L, 400L, 2137),
    error = function(e) conditionMessage(e)
  )
  expect_false(grepl("mcpower.app/report", err, fixed = TRUE))
})
